lib/auth/github.go

```diff
@@ -84,6 +91,89 @@ func (a *Server) upsertGithubConnector(ctx context.Context, connector types.Gith
 	return nil
 }
 
+// HTTPRequester allows a net/http.Client to be mocked for tests.
+type HTTPRequester interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// checkGithubFeatureSupport returns an error if any of the Github
+// organizations specified in this connector use external SSO. SSO
+// is a Teleport Enterprise feature that should not be allowed for OSS.
+// If client is nil a net/http.Client will be used.
+func checkGithubFeatureSupport(conn types.GithubConnector, client HTTPRequester) error {
+	version := modules.GetModules().BuildType()
+	if version == modules.BuildEnterprise {
+		return nil
+	}
+
+	if client == nil {
+		transport, err := defaults.Transport()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		client = &http.Client{
+			Transport: transport,
+		}
+	}
+
+	checkOrg := func(org string) error {
+		// A Github organization will have a "sso" page reachable if it
+		// supports external SSO. There doesn't seem to be any way to get this
+		// information from the Github REST API without being an owner of the
+		// Github organization, so check if this exists instead.
+		ssoURL := fmt.Sprintf("%s/%s/%s", githubOrgsURL, url.PathEscape(org), "sso")
+
+		var resp *http.Response
+		for i := 0; i < 3; i++ {
+			ctx, cancel := context.WithTimeout(context.Background(), defaults.HTTPRequestTimeout)
+			defer cancel()
+
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, ssoURL, nil)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			var urlErr *url.Error
+			resp, err = client.Do(req)
+			if err == nil {
+				break
+			} else if errors.As(err, &urlErr) && urlErr.Timeout() {
+				// Connection timed out, try to make the request again
+				continue
+			}
+			// Unknown error, don't try making any more requests
+			return trace.Wrap(err)
+		}
+		resp.Body.Close()
```

Comment:
It looks like `resp` will be `nil` if we time out 3x, need to handle that case and should probably add a test for it.

Reply:
Thanks, done